languages/tree-sitter-stack-graphs-java/src/stack-graphs.tsg

```diff
@@ -355,46 +866,136 @@ attribute node_symbol = node            => symbol = (source-text node), source_n
-    name: (identifier) @identifier
-    value: (object_creation_expression
-      type: (type_identifier) @object)
-  )) @local_variable {
-  node @object.value
-  attr (@object.value) node_reference = @object
-  node @type.value
-  attr(@type.value) node_reference = @type
-  edge @type.value -> @local_variable.after_scope
-  edge @local_variable.after_scope -> @object.value
-  edge @object.value -> @local_variable.after_scope
+(method_reference . (_) @lhs) @this {
+  edge @lhs.lexical_scope -> @this.lexical_scope
+}
+
+(method_reference . (_) @lhs (identifier) @id) @this {
+  node @this.member
+
+  node expr_ref
+  attr (expr_ref) node_reference = @id
+  edge expr_ref -> @this.member
+
+  attr (@this.member) push_symbol = "."
+  edge @this.member -> @lhs.type
+
+  node expr_ref__typeof
+  ; edge @this.type -> expr_ref__typeof ; FIXME: function types
+
+  attr (expr_ref__typeof) push_symbol = ":"
+  edge expr_ref__typeof -> expr_ref
+}
+
+(method_reference . (identifier) @lhs (identifier) @id) @this {
+  ; @lhs could be a type name
+  node ref
+  attr (ref) node_reference = @lhs
+  edge @this.member -> ref
+  edge ref -> @this.lexical_scope
+}
+
+(parenthesized_expression (_) @child) @expr {
+  edge @child.lexical_scope -> @expr.lexical_scope
 }
 
 ;; =============
 ;;  Expressions
 ;; =============
 
 [
-  ; FIXME: all expressions should be here
-  (assignment_expression)
+  (array_initializer)
+  (expression)
   (field_access)
-  (identifier)
-  (method_invocation)
+  (parenthesized_expression)
+  (primary_expression)
   (super)
-  (this)
 ] @expr
 {
   node @expr.type
   node @expr.lexical_scope
 }
 
+(assignment_expression left: (identifier) @ident) @this {
+  ; FIXME: can we combine this with primary_expression/identifier?
+  node @ident.lexical_scope
+  node @ident.type
+  node member
+  node implicit_this
+  node implicit_this__typeof
+
+  node object__typeof
+
+  attr (implicit_this) symbol_reference = "this", empty_source_span
+  edge implicit_this__typeof -> implicit_this
+  attr (implicit_this__typeof) push_symbol = ":"
+  edge implicit_this -> @ident.lexical_scope
+
+  node ref
+  attr (ref) node_reference = @ident
+  edge ref -> @ident.lexical_scope
+  attr (ref -> @ident.lexical_scope) precedence = 1
+  edge ref -> member
+
+  attr (member) push_symbol = "."
+
+  edge member -> object__typeof
+  edge object__typeof -> implicit_this__typeof
+
+  node ref__typeof
+  edge @ident.type -> ref__typeof
+
+  attr (ref__typeof) push_symbol = ":"
+  edge ref__typeof -> ref
+
+  edge @ident.lexical_scope -> @this.lexical_scope
+}
+
+(assignment_expression left: (field_access) @access) @this {
+  edge @access.lexical_scope -> @this.lexical_scope
+}
+
+(assignment_expression right: (_) @right) @this {
+  edge @right.lexical_scope -> @this.lexical_scope
+}
+
+(binary_expression left: (_) @lhs right: (_) @rhs) @this {
+  edge @lhs.lexical_scope -> @this.lexical_scope
+  edge @rhs.lexical_scope -> @this.lexical_scope
+}
```

Comment:
I feel like this scope propagation could usefully be some sort of default behaviour, or we could just carry around a dynamically scoped `lexical_scope` variable (😁), with binders overriding the default (or rebinding `lexical_scope`) and everything else just … not.

Reply:
There's a lot of really nice stuff you can do when you have a grammar oriented towards analysis, but I suspect that it would be hard to actually get something like that to work with a tree-sitter grammar. If we had some means of encoding in the grammar what the functional/semantic role of the parts was, this would be much easier. Alas.

Reply:
I think we could get by with passing the current lexical scope around in tsg, like if the rules all ran in `ReaderT <whatever>`.